core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -159,18 +160,30 @@ public void commit(TableMetadata base, TableMetadata metadata) {
     int nextVersion = (current.first() != null ? current.first() : 0) + 1;
     Path finalMetadataFile = metadataFilePath(nextVersion, codec);
     FileSystem fs = getFileSystem(tempMetadataFile, conf);
-
-    // this rename operation is the atomic commit operation
-    renameToFinal(fs, tempMetadataFile, finalMetadataFile, nextVersion);
-
-    LOG.info("Committed a new metadata file {}", finalMetadataFile);
-
-    // update the best-effort version pointer
-    writeVersionHint(nextVersion);
-
-    deleteRemovedMetadataFiles(base, metadata);
-
-    this.shouldRefresh = true;
+    boolean versionCommitSuccess = false;
+    try {
+      fs.delete(versionHintFile(), false /* recursive delete*/);
```

Comment:
By doing a delete here, and we write at line 175, I'm pretty sure that we lose atomicity. The file will be missing for a (very short) period of time.

Reply:
@Fokko 
Sir, the purpose of versionHintFile is simply to speed up the reading of the latest version. 

Its presence or absence does not affect the client's ability to correctly parse the latest version. In other words, it's just a sugar. When the versionHintFile doesn't exist, we traverse all the metadataFiles to find out what the latest version is. 

This logic is already implemented in hadoopcatalog. This means that our first task is to make sure that fs.rename(tempMetadataFile,finalMetadataFile) is executed successfully. As long as it does, the rest of the problem is trivial. 

Also, based on my discussions with Russell, we agreed that deleting the versionHintFile first would be a better option.
https://github.com/apache/iceberg/pull/9546#pullrequestreview-1868256978

Reply:
@Fokko 
In addition, actively deleting the versionHintFile not only improves data consistency between the versionHintFile and the actual latest version, it also weakens the use of the versionHintFile. Since we using a posix-style-filesystem, we have no way to atomically manipulate two or more files. Deprecating versionHintFile is an obvious solution. Once we have weakened versionHintFile to a certain point, we can deprecate it.

Reply:
Also, you may have noticed the introduction of lockManager in hadoopcatalog(discuss: https://apache-iceberg.slack.com/archives/C025PH0G1D4/p1647950504955309), but it doesn't really solve the problem, regardless of its implementation. LockManager only solves the problem of concurrent commits from multiple clients, but it doesn't solve the problem of commits being in an intermediate state due to exceptions, which may result in incorrect behaviour. 

In a file system that is basically posix compliant, we theoretically don't need lockManager to prevent multiple clients from committing concurrently, because fs.rename can do that. 

In an object store, we do need the lockManager, but we have some other issues to address.

Reply:
> Sir, the purpose of versionHintFile is simply to speed up the reading of the latest version. it‘s just a index file.

That is correct, but it is also highly coupled to the Iceberg project since Iceberg optimizes for object stores. Object stores are terrible at listing files since they are often paged responses. These are both slow and costly, therefore the pointer in the `versionHintFile` makes this much faster.

> In a file system that is basically posix compliant, we theoretically don't need lockManager to prevent multiple clients from committing concurrently, because fs.rename can do that.

How would that work? I think a race condition cannot be avoided, where it would pass 99.99% of the time, but it can be that they overwrite each other because they don't know that multiple processes are writing to the table.

Reply:
@Fokko 
>How would that work? I think a race condition cannot be avoided, where it would pass 99.99% of the time, but it can be that they overwrite each other because they don't know that multiple processes are writing to the table.

For filesystems that support atomic operations, rename operations do not overwrite each other. On the other hand, this operation is not supported in object storage. Due to the atomic nature of the operation, each client will know if its operation succeeded or not when it renames.

Reply:
> That is correct, but it is also highly coupled to the Iceberg project since Iceberg optimizes for object stores. Object stores are terrible at listing files since they are often paged responses. These are both slow and costly, therefore the pointer in the `versionHintFile` makes this much faster.

@Fokko 
For object storage, since the current implementation of fileSystemCatalog is based on the assumption that the file system supports atomic operations, the current implementation can not be used for object storage, either we provide another implementation, or we use some CNCF's middleware to proxy, which usually can provide atomic operation features.